enhancements/hybrid-helm-operator.md

```diff
@@ -0,0 +1,223 @@
+  the extensibility of Golang-based operators.
+
+### Non-Goals
+
+- We will not continue to manage the hybrid operator in any way, once it is converted
+  it is essentially a Go-based operator and so will not make use of a base image with
+  dependencies, automatic upgrades of helm versions, or any other conveniences provided
+  by the base image. The user will need to manage and upgrade their code by hand.
+
+## Proposal
+
+There are 3 stages to this proposal:
+
+1. Joe Lanford has made significant progress on this already, in the
+   [joelanford/helm-operator][helm-operator-repo] repository.
+   First we will migrate this code to the operator-sdk repo, where it will replace the
+   existing helm-operator code.
+1. We will write a new SDK plugin that can scaffold out a hybrid operator. As a first
+   iteration we can just scaffold out most of the code that is currently in the
+   [helm-operator][helm-operator-repo] (though the exact structure of the code will likely change)
+   and use that to generate what we ship in the base image.
+1. We will break as much code as we reasonably can out of the scaffolding and into a
+   utility library, to minimize the amount of code the user will need to maintain. This
+   library could either be the existing operator-lib package, or a new, helm-operator specific
+   package.
+
+### User Stories
+
+#### Story 1
+
+As a Helm-based operator author, I want to add additional programmatic backup and restore
+functionality to my operator, without starting over with a completely different type of operator.
+
+#### Story 2
+
+As a Helm-based operator author, I want to change core behavior of the reconciliation loop
+that is not currently exposed in the watches.yaml
+
+#### Story 3
+
+As a prospective author of a Helm-based operator, I want to know that if I invest time and
+effort creating a Helm-based operator, I won't need to start over in order to create a higher
+level operator.
+
+#### Story 4
+
+As a Golang-based operator author, I would like to leverage existing helm charts for deploying
+pieces of my application, without sacrificing the flexibility of Go.
+
+
+### Implementation Details/Notes/Constraints
+
+#### Scaffolding plugin for the SDK
+
+The basic idea behind the hybrid Helm/Go operator plugin is that we will scaffold out a Go
+operator project structure, with all of the code bits from the helm-operator project filled in,
+such that building the project without adding any APIs or modifying any code will result in the
+same container as what is shipped in the helm-operator base image.
+
+Because the driving need behind this functionality is the ability to customize any piece of
+the Helm-based operator, we will scaffold out every bit of code that is not accessible via
+a library, and err on the side of scaffolding more than we need to.
+
+For the first iteration of the hybrid Helm/Go operator plugin, we will likely just scaffold
```

Comment:
My original idea of what to scaffold was something along the lines of what's in this file: https://github.com/joelanford/helm-operator/blob/main/internal/cmd/run/cmd.go#L99

Reply:
@fabianvf could you expand on the pros and cons of scaffolding out basically all of helm-operator's existing libraries in user projects vs scaffolding code that imports from it.

I (perhaps naively) considered the library bits to be pretty foundational building blocks, and if folks find they can't do something with one particular piece of the library, they can copy and maintain just that piece (or even better, submit a PR to the library with whatever new feature or bugfix they need).

The pretty obvious benefit of delivery via library is users get bug fixes and features for essentially free (just update your go.mod file).

The pretty obvious benefit of copying all the code into a user project is that they have ultimate control and absolutely zero hurdles to change things as necessary.

One wrinkle of maintaining all the code in the scaffold is that it _may_ be harder to maintain it. Though perhaps the go1.16 embedding feature can be used to have the library code on disk as normal library files and then a separate package can embed all of it for use in a scaffolding plugin.

Reply:
Yeah I think I should update the phrasing here, I wrote this bit earlier in the process. I think there's quite a lot of code in the `helm-operator` that definitely should go straight to a library, like the `annotations`, `client`, `hook`, `manifestutil`, and `values`, which are either made up of foundational code or export easily implementable interfaces if users need to modify their behavior. I think the main bits that users have been wanting to modify are the main, reconciler and configuration bits. I'll narrow the language here.

Reply:
I can definitely see a argument for the contents of reconciler itself being scaffolded. 

If we do that, then the hooks and value mapper become less valuable because they could essentially be ripped out and replaced with anything a user wants. That's not a bad thing in my mind, just a different (and more flexible way) for users to approach their problems.

I think the one big tradeoff is still the free bug fixes and features, but perhaps we could have a `create api` flag for this plugin like `--scaffold-reconciler` that lets a user choose between using the library or getting the code dumped into their project.

Reply:
Yeah, and of course we can also move many of the granular bits of logic out of the reconciler and into the library, plus with this proposal the amount of code scaffolded in phase 1 will be slimmed down significantly before it ever hits a stable/supported state. I'm just trying to err on the side of flexibility in the initial implementation, so that we can fine tune it based on what people are actually changing/using